api/ai/algorithm_runner.py

```diff
@@ -52,12 +54,25 @@ def __init__(
     def generate(self, students: List[Student]) -> TeamSet:
         # the algorithm classes internally track generated teams, so a new instance of the
         #   algorithm class MUST be created to run a new generation without side effects
+        # This generated instance is initialized in the prepare method
+        return self.algorithm.generate(students)
+
+    def prepare(self, students: List[Student]) -> None:
+        """
+        Find the algorithm class and run prepare if possible
+        """
         algorithm = self.algorithm_cls(
             team_generation_options=self.team_generation_options,
             algorithm_options=self.algorithm_options,
             algorithm_config=self.algorithm_config,
         )
-        return algorithm.generate(students)
+        self.algorithm = algorithm
+
+        try:
+            algorithm.prepare(students)
+        except NotImplementedError:
```

Comment:
Idk if I like this, just have the prepare method on the abstract class `pass` instead of throwing not implemented. It's not something we need every algorithm to have, and it is acceptable for the ones that don't need to just do nothing at that step